### PR TITLE
Introduce baseAdapter capability, add support for multiple adapter configurations

### PR DESCRIPTION
**Pull request by webflo**
I tried to use gush with gitlab.com and our self-hosted corporate gitlab installation simultaneously, but the adapter config is keyed by adapter name. This means the two adapter settings override each other.

This PR add the concept of baseAdapter and configured adapter. This enables multiple adapter configurations. I think is also possible to replace the GitHub Enterprise and Jira Enterprise adapter with concept.

Example configuration in gush.yml

```
# Gush configuration file, any comments will be lost.
repo_adapter: gitlab:gitlab.mycompany.com
issue_tracker: gitlab:gitlab.mycompany.com
repo_org: foo
repo_name: bar
issue_project_org: foo
issue_project_name: bar
```

I opened a PR for the gitlab adapter. https://github.com/gushphp/gush-gitlab-adapter/pull/18

### CODE REVIEW

**Comment by webflo** [comment]
@cordoval @sstok I would like to get some feedback, before i continue to work on it add test coverage etc.

**Comment by sstok** [comment]
I actually had a good idea for this :) https://github.com/gushphp/gush/issues/401 (see Enterprise support). Just need to find the time (and energy (stupid allergies)) to get working on this.

If you can adapt my original idea, I'm willing to help you with filling in the blanks to finally making this possible :+1: good to see you using Gush :smile:

File: src/Factory/AdapterFactory.php [hunk 1]
```diff
@@ -224,4 +231,10 @@ private function guardFactoryClassImplementation($adapterFactory, $label)
             self::SUPPORT_ISSUE_TRACKER => $issueTracker,
         ];
     }
+
+    private  function getBaseAdapter($name) {
+      $array = explode(':', $name);
+      return $array[0];
+    }
+
```
Comment (cordoval) [inline src/Factory/AdapterFactory.php]: remove this line break,

also @sstok what are your thoughts on this one?

Comment (sstok) [inline src/Factory/AdapterFactory.php]: Only the GIT url will not work as the API and web URL also need to be configured (see mentioned issue).

Comment (webflo) [inline src/Factory/AdapterFactory.php]: Yes thats true, but i wanted to keep the patch small because currently the URL for the gitlab adapter is stored in ``~/.gush/.gush.yml`` and not the repo.

I think it makes more sense to store it in the repo itself, but it will break BC ...

Comment (cordoval) [inline src/Factory/AdapterFactory.php]: we are not concerned about BC at all, gush is a tool and we can break BC anytime needed.

At least for now, it is stable but if this is needed we can do it without problems, ping @sstok

**Comment by cordoval** [comment]
good job so far @webflo :+1:

File: src/Adapter/DefaultConfigurator.php [hunk 2]
```diff
@@ -185,4 +190,34 @@ public function validateUrl($url)
...
+    {
+        return $this->baseAdapterName;
+    }
+
+    /**
+     * Set the base adapter name.
+     *
+     * @return string
+     */
+    public function setBaseAdapterName($name)
```
Comment (sstok) [inline src/Adapter/DefaultConfigurator.php]: This can be moved to the constructor as a configurator should never change its actual adapter base-name.
Setters to require something is set (without it the class will not work) will the make the class more prone to bugs.

Besides that :+1: for this change.

**Comment by cordoval** [comment]
@webflo what is blocking further progress on this

**Comment by cordoval** [comment]
bump @sstok we may have to continue this effort, i cannot recall what this can bring to the table

**Comment by webflo** [comment]
@cordoval Sorry a had a few other things to do. I worked on it yesterday evening. I update the PR soon.

**Comment by webflo** [comment]
I had to introduce new commandline parameters. repo_adapter_config and issue_tracker_config. I tried to store everything in repo_adapter, but it would take a lot of refactoring.

```
repo_adapter: gitlab
repo_adapter_config:
  base_url: http://example.com/api/v3
  domain_url: http://example.com/api/v3
issue_tracker: gitlab
issue_tracker_config:
  base_url: http://example.com/api/v3
  domain_url: http://example.com/api/v3
repo_org: foo
repo_name: bar
issue_project_org: foo
issue_project_name: bar
```

**Comment by cordoval** [comment]
@webflo add me on skype cordova_luis

**Comment by cordoval** [comment]
@sstok i remember i used github and githubenterprise at the same time way back

i am not following what could have happened that this was not possible anymore
hmm they had different classes indeed

File: src/Factory/LegacyAdapterFactory.php [hunk 3]
```diff
@@ -77,4 +77,5 @@ public function getClassName()
     {
         return $this->className;
     }
+
```
Comment (sstok) [inline src/Factory/LegacyAdapterFactory.php]: Please revert this change.

**Comment by sstok** [comment]
@webflo can you fix the failing tests?

**Comment by cordoval** [comment]
bump @webflo

**Comment by cordoval** [comment]
I am going to be using gush with gitlab as well cc @sstok do you foresee this same problem?

**Comment by sstok** [comment]
@cordoval It depends on if you are using Gitlab.com or a private hosted instance.
In the second case it will only work partially, as you can only configure one.

I will see what I can do once RollerworksSearch is completed ;)

**Comment by sstok** [comment]
I'm taking over this PR (with original authorship preserved ;) ) thank you @webflo for the work done so far.

**Comment by webflo** [comment]
@sstok Thanks for working on it.

**Comment by cordoval** [comment]
he is da mang :+1: :blush:

File: src/Factory/AdapterFactory.php [hunk 4]
```diff
@@ -218,10 +218,12 @@ private function guardFactoryClassImplementation($adapterFactory, $label)
...
         return [
-            'factory' => $adapterFactory,
+            'name' => $name,
             'label' => $label,
+            'factory' => $adapterFactory,
             self::SUPPORT_REPOSITORY_MANAGER => $repositoryManager,
             self::SUPPORT_ISSUE_TRACKER => $issueTracker,
         ];
     }
+
```
Comment (GrahamCampbell) [inline src/Factory/AdapterFactory.php]: ?

File: src/Util/ConfigUtil.php [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ */
+
+namespace Gush\Util;
+
+class ConfigUtil {
+
+  public static function generateConfigurationIdentifier($adapter, array $adapterConfig) {
+    return $adapter . ':' . substr(sha1(serialize($adapterConfig)), 0, 8);
+  }
+
```
Comment (GrahamCampbell) [inline src/Util/ConfigUtil.php]: ?

Comment (sstok) [inline src/Util/ConfigUtil.php]: CS issues are addressed in my branch, just need some time to change some things to not introduce the many BC breaks ;)

**Comment by sstok** [comment]
Depends on #553 to be fully workable, I will try to finish this asap.

**Comment by cordoval** [comment]
bump @webflo last call on this one, it seems it is too adhoc and not a common use case so 👎

**Comment by sstok** [comment]
I will try to finish this as soon as possible, but I also have a ton of other things to do.

Some notes:

```
* Allow old Gush configuration format.
* Provide defaults for the api and repository URI, unless given explicitly
* Internally the a configuration hash-key is used.
    * When no configuration exists for the current combination the configurator is activated and the settings are saved.
* Allow to provide the configuration using the CLI options, add the new `—api-uri` and `—repo-uri` options to provide this.
* The adapter-name is only used for the registry
* Configuring an adapter, will configure the ‘default’ adapter; not actually called ‘default’ but it uses the default values.
* Dump file after successful authentication (to ease the process, not at the end when authentication keeps failing).

Fully backward compatible and still supports a new adapter format!

core:configure:

* Show adapters list
    * select adapter
        * Show configurations (plus add)
            * Select configuration
                * Option: remove, reconfigure
```

**Comment by cordoval** [comment]
@sstok don't worry we will keep this in mind for future refactors.
